core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerde.java

```diff
@@ -214,23 +220,17 @@ public Page deserialize(PagesSerdeContext context, Slice serializedPage)
         return readRawPage(positionCount, slice.getInput(), blockEncodingSerde);
     }
 
-    public static Slice readSerializedPage(SliceInput input)
+    public static Slice readSerializedPage(Slice headerSlice, InputStream inputStream)
+            throws IOException
     {
-        int positionCount = input.readInt();
-        byte marker = input.readByte();
-        int uncompressedSize = input.readInt();
-        int compressedSize = input.readInt();
-
-        SliceOutput output = Slices.allocate(SERIALIZED_PAGE_HEADER_SIZE + compressedSize).getOutput();
-        output.writeInt(positionCount);
-        output.writeByte(marker);
-        output.writeInt(uncompressedSize);
-        output.writeInt(compressedSize);
-
-        Slice result = output.getUnderlyingSlice();
-        input.readBytes(result, SERIALIZED_PAGE_HEADER_SIZE, compressedSize);
-
-        return result;
+        checkArgument(headerSlice.length() == SERIALIZED_PAGE_HEADER_SIZE,
+                "headerSlice length should equal to %s", SERIALIZED_PAGE_HEADER_SIZE);
+
+        int compressedSize = getIntUnchecked(headerSlice, COMPRESSED_SIZE_OFFSET);
+        byte[] outputBuffer = new byte[SERIALIZED_PAGE_HEADER_SIZE + compressedSize];
+        arraycopy(headerSlice.byteArray(), 0, outputBuffer, 0, SERIALIZED_PAGE_HEADER_SIZE);
```

Comment:
Minor: since we’re using headerSlice.byteArray() here but the Slice is created externally, it would be safer to use headerSlice.byteArrayOffset() instead of assuming that the offset is 0

Reply:
Actually I wonder if I should just use `headerSlice.getBytes(0, outputBuffer);`. Do you think this makes any difference in performance?

Reply:
It shouldn't matter much here, but I suspect `System.arraycopy` would have a slight edge over `Slice#getBytes` because of the way that `Slice` uses `Unsafe.copyMemory`

Reply:
I find https://groups.google.com/g/mechanical-sympathy/c/sug91A1ynF4 saying that `Unsafe.copyMemory` is slight faster than `System.arraycopy`. So using `headerSlice.getBytes(0, outputBuffer, 0, SERIALIZED_PAGE_HEADER_SIZE);` instead.